### PR TITLE
Parallel parser

### PR DESCRIPTION
**Pull request by TobiasMeisel**
- logs from all MPI processes are now considered
- simplified logic - all occurrences of a known regex will be added to a
list of records
- examples for further processing based on pandas, see unit tests / ogs_common_analyses: 
  - convergence analysis of newton iteration
  - convergence analysis of coupling iteration in staggered scheme
  - time step analysis

### CODE REVIEW

File: ogs6py/log_parser/common_ogs_analyses.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+    pt = dfe_convergence_coupling_iteration.pivot_table(interest, context)
+    check_output(pt, interest, context)
+    return pt
+
+
+def time_step_vs_iterations(df):
+    interest = 'iteration_number'
+    context = 'time_step'
+    check_input(df, interest, context)
```
Comment (joergbuchwald) [inline ogs6py/log_parser/common_ogs_analyses.py:118]: This line is causing some trouble. It looks like it is taking every letter as a column.

**Comment by joergbuchwald** [comment]
I also tested a some bigger logfiles: It took me 30GB RAM for processing, but I was finally able to open a 2GB log file within 10 min.
Not sure, whether it was the same for the old parser, but it reminds me that processing a lot of these files in parallel could be quite a challenge on some machines. Perhaps, we should think about a mechanism that allows to collect only selected data to save some memory, instead of collecting everything and filtering afterwards.

File: ogs6py/log_parser/common_ogs_analyses.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    check_output(pt, interest, context)
+    return pt
+
+
+def analysis_convergence_newton_iteration(df):
+    dfe_newton_iteration = df.copy()
+    interest = ['dx', 'x', 'dx_x']
+    if 'coupling_iteration' in df:
+        context = ['time_step', 'coupling_iteration', 'process',
+                   'iteration_number', 'component']
```
Comment (joergbuchwald) [inline ogs6py/log_parser/common_ogs_analyses.py:74]: I think the 'component' keyword optional, it is only there if a multicomponent variable like displacement is involved.

File: ogs6py/log_parser/common_ogs_analyses.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+        check_input(df, interest, context)
+        pt = dfe_newton_iteration.pivot_table(interest, context)
+
+    check_output(pt, interest, context)
+    return pt
+
+
+@pre_post_check(interest=['dx', 'x', 'dx_x'], \
+                context=['time_step', 'coupling_iteration', 'coupling_iteration_process',
+                         'component'])
```
Comment (joergbuchwald) [inline ogs6py/log_parser/common_ogs_analyses.py:95]: As above. 'component' should be optional.

**Comment by TobiasMeisel** [comment]
> I also tested a some bigger logfiles: It took me 30GB RAM for processing, but I was finally able to open a 2GB log file within 10 min. Not sure, whether it was the same for the old parser, but it reminds me that processing a lot of these files in parallel could be quite a challenge on some machines.

Interesting. RAM/DISK_STORAGE factor seems to be too high. It is in part a consequence of duplication of data (e.g. records and the pandas dataframe). This problem needs special focus, and would suggest that it will be not part of this pull request. But I believe that the new design allows all parts to be run in parallel and that there will be a solution for reducing memory consumption as well.

> Perhaps, we should think about a mechanism that allows to collect only selected data to save some memory, instead of collecting everything and filtering afterwards.

If we had the optional parameter for ogs regexes, we would get with a reduced list of regexes 1. Less memory consumption(smaller records list) and 2. Save some time in the inner loop (iterating over all regexes)
```
def parse_file(file_name, maximum_lines=None, force_parallel=False,ogs_res = ogs_regexes()):
    
```

File: ogs6py/log_parser/common_ogs_analyses.py (hunk 2 above)
Comment (dominik-kern) [reply ogs6py/log_parser/common_ogs_analyses.py:74]: You are right, that the _component_ is uninteresting when you have scalar processes or at least scalar convergence criteria, but if it is not so, then it is helpful to have it. IMHO having one variable more does not justify to create a flag or another analysis, so I would not change this.

Comment (TobiasMeisel) [reply ogs6py/log_parser/common_ogs_analyses.py:74]: Agree with Dominik here in terms that we do not add another analysis. I would treat the functions in common_ogs_analysis as examples and good starting point for specific/custom analyses. In this special case here the study without component seems to be more generic and would follow Joergs proposal to remove component.

Comment (joergbuchwald) [reply ogs6py/log_parser/common_ogs_analyses.py:74]: We could check whether `'component'` is there in the data and add it conditionally to the context.

**Comment by joergbuchwald** [comment]
> > I also tested a some bigger logfiles: It took me 30GB RAM for processing, but I was finally able to open a 2GB log file within 10 min. Not sure, whether it was the same for the old parser, but it reminds me that processing a lot of these files in parallel could be quite a challenge on some machines.
> 
> Interesting. RAM/DISK_STORAGE factor seems to be too high. It is in part a consequence of duplication of data (e.g. records and the pandas dataframe). This problem needs special focus, and would suggest that it will be not part of this pull request. But I believe that the new design allows all parts to be run in parallel and that there will be a solution for reducing memory consumption as well.
> 
> > Perhaps, we should think about a mechanism that allows to collect only selected data to save some memory, instead of collecting everything and filtering afterwards.
> 
> If we had the optional parameter for ogs regexes, we would get with a reduced list of regexes 1. Less memory consumption(smaller records list) and 2. Save some time in the inner loop (iterating over all regexes)
> 
> ```
> def parse_file(file_name, maximum_lines=None, force_parallel=False,ogs_res = ogs_regexes()):
>     
> ```

That could be a ways to do it, yes. I was originally thinking of a list with regexes to omit, but at the end it  does not make things easier.

File: ogs6py/log_parser/common_ogs_analyses.py (hunk 3 above)
Comment (joergbuchwald) [reply ogs6py/log_parser/common_ogs_analyses.py]: I gues, I fixed it now.

**Pull request merged**